### PR TITLE
Fix multi-tensor LAMB reduction to be deterministic

### PR DESCRIPTION
**Pull request by suffiank**
**Description**: Fix multi-tensor LAMB reduction to be deterministic by ordering reduction across thread blocks.

image

Sister PR to https://github.com/microsoft/onnxruntime/pull/6023 (Only one to be checked in.)

**Motivation and Context**
- Why is this change required? What problem does it solve?
Makes LAMB reduction kernel deterministic.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 1]
```diff
@@ -10,6 +16,18 @@
...
+__forceinline__ __host__ __device__ int least_pow2_bound(int value) {
+  unsigned int value_ = static_cast<unsigned int>(value);
+  --value_;
+  value_ |= value_ >> 1;
+  value_ |= value_ >> 2;
+  value_ |= value_ >> 4;
+  value_ |= value_ >> 8;
+  value_ |= value_ >> 16;
+  return static_cast<int>(++value_);
+}
```
Comment (KeDengMS) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:29]: maybe making use of some CUDA integer intrinsics?
[https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html#group__CUDA__MATH__INTRINSIC__INT](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html#group__CUDA__MATH__INTRINSIC__INT)

**Review by wschin — DISMISSED** [review]

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 2]
```diff
@@ -442,7 +460,7 @@ INSTANTIATE_LAMB_MULTI_TENSOR_UPDATE_FUNCTOR(float, float, half, half)
 
 template <typename TIn1, typename TIn2, typename TOut1, typename TOut2, typename TBuf>
 __launch_bounds__(ChunkGroup<4>::thread_count_per_block)
-__global__ void LambMultiTensorReductionImpl(ChunkGroup<4> chunk_group) {
+__global__ void LambMultiTensorReductionImpl(ChunkGroup<4> chunk_group, TOut1* w_buffer, TOut2* d_buffer) {
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:463]:
```suggestion
// w_buffer[i] is used to store the squared sum of all elements processed by the i-th block.
// d_buffer[i] ...
__global__ void LambMultiTensorReductionImpl(ChunkGroup<4> chunk_group, TOut1* w_buffer, TOut2* d_buffer) {
```

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 1 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:29]: It has been defined in `/bert_ort/wechi/onnxruntime1/onnxruntime/core/providers/cuda/reduction/reduction_utils.cuh`

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 2 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:453]: Yeah, agree. Added comment.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 1 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu]: Removed this code snippet and added the appropriate header.

Comment (KeDengMS) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu]: Don't have to be in this PR, but maybe a chance to improve in reduction_utils.cuh

Comment (KeDengMS) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu]: Maybe this: ```return 1 << (32 - __clz(value_ - 1));```

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 3]
```diff
@@ -519,11 +577,20 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
   ORT_ENFORCE(thread_count % GPU_WARP_SIZE == 0);
   ORT_ENFORCE((thread_count & (thread_count - 1)) == 0);
 
-  LambMultiTensorReductionImpl<TIn1, TIn2, TOut1, TOut2, TBuf><<<chunk_group.chunk_count, thread_count, shared_memory_size>>>(chunk_group);
+  const int num_blocks = chunk_group.chunk_count;
+  size_t w_buffer_size = static_cast<int>(num_blocks * sizeof(TOut1) + sizeof(int));
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:581]: why static_cast<int> and assign to size_t?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc [hunk 4]
```diff
@@ -543,9 +545,15 @@ Status LambOptimizer<T1, T2, T3, T4, T_GRAD_NORM, T_MIXED_PRECISION_FP>::Compute
...
 
   // Allocate a buffer in byte for reduction API calls.
-  const auto reduction_buffer_size =
+  auto reduction_buffer_size =
       compute_reduction_buffer_size<CudaT2>(max_tensor_size);
 
+  // Enlarge reduction buffer to accomodate multi-tensor reduction kernel as well
+  const int tensor_group_size = 4; // w, d, w_norm, d_norm
+  const int max_blocks = ChunkGroup<tensor_group_size>::max_block_count;
+  const size_t multitensor_buffer_size = sizeof(CudaT2)*static_cast<int>(max_blocks*sizeof(CudaT2) + sizeof(int));
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:554]: why is the `static_cast<int>` needed?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc [hunk 5]
```diff
@@ -543,9 +545,15 @@ Status LambOptimizer<T1, T2, T3, T4, T_GRAD_NORM, T_MIXED_PRECISION_FP>::Compute
   }
 
   // Allocate a buffer in byte for reduction API calls.
-  const auto reduction_buffer_size =
+  auto reduction_buffer_size =
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:548]: fyi, a trick for keeping things const:
```
const auto x = [&](){
  ...
  return value;
}();
```

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 6]
```diff
@@ -1,9 +1,16 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/providers/cuda/cuda_allocator.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:4]: nit: order of includes
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 7]
```diff
@@ -519,11 +577,20 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
   ORT_ENFORCE(thread_count % GPU_WARP_SIZE == 0);
   ORT_ENFORCE((thread_count & (thread_count - 1)) == 0);
 
-  LambMultiTensorReductionImpl<TIn1, TIn2, TOut1, TOut2, TBuf><<<chunk_group.chunk_count, thread_count, shared_memory_size>>>(chunk_group);
+  const int num_blocks = chunk_group.chunk_count;
+  size_t w_buffer_size = static_cast<int>(num_blocks * sizeof(TOut1) + sizeof(int));
+  size_t d_buffer_size = static_cast<int>(num_blocks * sizeof(TOut2) + sizeof(int));
+
+  ORT_ENFORCE(w_buffer_size*sizeof(TOut1) + d_buffer_size*sizeof(TOut2) <= reduction_buffer_size);
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:584]: maybe i'm missing something - it seems like w/d_buffer_size are in bytes already, why do we multiply by sizeof(TOut1/2) here?
also, i guess w_buffer_size has space for an extra int for use as the block completion counter. why does d_buffer_size also include an extra int?
do we need to account for alignment of TOut1's, TOut2's and the int?

**Comment by suffiank** [comment]
@edgchen1 I had to make significant correction after realizing the block reduction should only take place for blocks corresponding to the same tensor. Some of your comments got fixed in relation to that. I am going to redo the perf analysis and post it.

**Comment by suffiank** [comment]
No significant perf change after code update.
image

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc [hunk 8]
```diff
@@ -267,6 +267,7 @@ Status launch_lamb_compute_direction(
 
 template <typename CudaTNorm, typename CudaTIn1, typename CudaTIn2>
 Status launch_lamb_reduction(
+    const CudaKernel* kernel,
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:270]: >const CudaKernel* [](start = 4, length = 17)

can this be `const CudaKernel&` so it's not possible to have a null value?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 9]
```diff
@@ -440,9 +443,15 @@ INSTANTIATE_LAMB_MULTI_TENSOR_UPDATE_FUNCTOR(double, double, double, half)
 INSTANTIATE_LAMB_MULTI_TENSOR_UPDATE_FUNCTOR(half, float, half, half)
 INSTANTIATE_LAMB_MULTI_TENSOR_UPDATE_FUNCTOR(float, float, half, half)
 
+// w_buffer[i], d_buffer[i] is used to store the squared sum of all elements processed by the i-th block.
+// sync_range_and_lock[t] contains (start block, number of blocks, locking counter) for tensor t
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:447]: >(start block, number of blocks, locking counter) [](start = 35, length = 48)

could it just be a struct? then the fields can have names

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 6 above)
Comment (edgchen1) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu]: "orttraining/training_ops/cuda/optimizer/lamb.h" should be first

---
In reply to: [538969572](https://github.com/microsoft/onnxruntime/pull/6028#discussion_r538969572) [](ancestors = 538969572)

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc (hunk 8 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:270]: Well, it's an upcast from LambOptimizer* to CudaKernel*.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 9 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:447]: Yeah, good point.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc (hunk 8 above)
Comment (edgchen1) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:270]: that should work for references too

---
In reply to: [539740511](https://github.com/microsoft/onnxruntime/pull/6028#discussion_r539740511) [](ancestors = 539740511)

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 9 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu]: OK, should read easier now.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc (hunk 8 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:270]: Ah, OK. I didn't realize that.

I gave it a try converting '*' to '&' in signatures but got an error I'm unsure on
```
/nfs/sukha/onnxruntime/onnxruntime/core/providers/cuda/multi_tensor/common.cuh:75:6: error: 
invalid abstract parameter type ‘onnxruntime::cuda::CudaKernel’
```

Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc]: Resolved using perfect forwarding. Thanks for the help!

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 10]
```diff
@@ -519,11 +572,34 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
...
+  const int max_tensors = ChunkGroup<4>::max_tensor_group_count;
+  LambMultiTensorSyncRangeAndLock initial = {0, 0, 0};
+  CudaKernel::CudaAsyncBuffer<LambMultiTensorSyncRangeAndLock> sync_range_and_lock(&kernel, initial, max_tensors);
+  for (int block_index = num_blocks-1; block_index >= 0; block_index--) {
+    int tensor_index = chunk_group.block_index_to_tensor_group_index[block_index];
+    auto& tensor_block_span = sync_range_and_lock.CpuPtr()[tensor_index];
+    tensor_block_span.leading_block = block_index;
+    tensor_block_span.number_blocks++;
+  }
+  sync_range_and_lock.CopyToGpu();
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:595]: Let's wrap this place in a new function such as
```
CudaKernel::CudaAsyncBuffer<LambMultiTensorSyncRangeAndLock> compute_tensor_range_and_lock(...)
```

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 11]
```diff
@@ -519,11 +572,34 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
   ORT_ENFORCE(thread_count % GPU_WARP_SIZE == 0);
   ORT_ENFORCE((thread_count & (thread_count - 1)) == 0);
 
-  LambMultiTensorReductionImpl<TIn1, TIn2, TOut1, TOut2, TBuf><<<chunk_group.chunk_count, thread_count, shared_memory_size>>>(chunk_group);
+  const int num_blocks = chunk_group.chunk_count;
+  size_t w_buffer_size = num_blocks * sizeof(TOut1);
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:576]: >size_t [](start = 2, length = 6)

const size_t

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu [hunk 12]
```diff
@@ -519,11 +572,34 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
...
+  size_t d_buffer_size = num_blocks * sizeof(TOut2);
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:577]: >size_t [](start = 2, length = 6)

const size_t

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.h [hunk 13]
```diff
@@ -126,7 +126,17 @@ struct LambMultiTensorComputeDirectionFunctor {
...
 struct LambMultiTensorReductionFunctor {
-  void operator()(ChunkGroup<4> chunk_group);
+  void operator()(
+    ChunkGroup<4> chunk_group, 
+    const CudaKernel& kernel, 
+    void *reduction_buffer, 
+    size_t reduction_buffer_size);
+};
+
+struct LambMultiTensorSyncRangeAndLock {
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.h:136]: >LambMultiTensorSyncRangeAndLock [](start = 7, length = 31)

Need a document for this struct like others in this file.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc [hunk 14]
```diff
@@ -542,9 +544,19 @@ Status LambOptimizer<T1, T2, T3, T4, T_GRAD_NORM, T_MIXED_PRECISION_FP>::Compute
...
-      compute_reduction_buffer_size<CudaT2>(max_tensor_size);
+  const size_t reduction_buffer_size = [&](){
+
+    // Allocate a buffer in byte for reduction API calls.
+    size_t rbs = compute_reduction_buffer_size<CudaT2>(max_tensor_size);
+
+    // Enlarge reduction buffer to accomodate multi-tensor reduction kernel as well
+    const int tensor_group_size = 4; // w, d, w_norm, d_norm
+    const int max_blocks = ChunkGroup<tensor_group_size>::max_block_count;
+    const size_t multitensor_block_reduce_buffer_size = 2*max_blocks*sizeof(CudaT2);
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:555]: >* [](start = 57, length = 1)

Add spaces between operands? Maybe format this file?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu (hunk 10 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu:595]: Good idea.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc (hunk 14 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.cc:555]: Prefer not to change other parts of file, so just added spacing about operands.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.h (hunk 13 above)
Comment (suffiank) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.h:136]: OK, added. Tried to be clear. Should suffice I think.

**Review by wschin — APPROVED** [review]
:shipit:

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**